File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        # convert BST tree to a sorted array:
        def bstToArrayInorder(root, arr):
            if root is None:
                return
            bstToArrayInorder(root.left, arr)
            arr.append(root.val)
            bstToArrayInorder(root.right, arr)
            return
        
        def bstBalancedConstructerFromSortedArray(arr):
            if not arr:
                return None
            
            mid = len(arr) // 2

            root = TreeNode(arr[mid])
            root.left = bstBalancedConstructerFromSortedArray(arr[:mid])
            root.right = bstBalancedConstructerFromSortedArray(arr[mid+1:])
            
            return root

        arr = []
        bstToArrayInorder(root, arr)
        
        return bstBalancedConstructerFromSortedArray(arr)

File: 1382-balance-a-binary-search-tree/1382-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
    vector<TreeNode*> sorted;
public:
    void inorder(TreeNode* root) {
        if (root == NULL) 
            return;
        
        inorder(root->left);
        sorted.push_back(root);
        inorder(root->right);
    }
    
    TreeNode* createTree(int start, int end) {
        if (start > end) 
            return NULL;
        
        int mid = start + (end - start)/2;
        TreeNode *root = sorted[mid];
        root->left = createTree(start, mid - 1);
        root->right = createTree(mid + 1, end);
        return root;
        
    }
    TreeNode* balanceBST(TreeNode* root) {
        inorder(root);
        return createTree(0, sorted.size()-1);
    }
};